BUG: Have a single histogram threshold calculator for intermodes threshold

jhlegarreta opened the pull request:
Fix the threshold calculator inheritance issue in
`itk::IntermodesThresholdImageFilter` by having a single histogram
threshold calculator variable.

The class was declaring a histogram threshold calculator under the
`m_IntermodesCalculator` ivar, which was of type
`itk::IntermodesThresholdCalculator` type. At the same time, the class was
inheriting from `itk::HistogramThresholdImageFilter`, which was declaring
its own histogram threshod calculator `m_Calculator`, which was of type
`itk::HistogramThredholdCalculator`.

This was making the `itk::IntermodesThresholdImageFilter` objects to have
two histogram calculator instances, and given that the `GenerateData`
method was implemented in the parent class
(`itk::HistogramThresholdImageFilter`) and making use of the
`m_Calculator` ivar, this was yielding a runtime error.

Fixes the runtime exception error mentoned in [this
comment](https://github.com/InsightSoftwareConsortium/ITK/pull/511#issuecomment-497877457).

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [X] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKExamples) for all new major features (if any)

jhlegarreta commented:
This is the approach I came up with to solve the issue raised in the comment referenced in the commit message.

I fear the current `ITKThresholding` module tests will not allow to evidence what this patch set tries to address, but if one cherry-picks the commit into PR #511, and makes the necessary adjustments (the `expectedThredshold` parameter tests are reached now and the baseline images can be generated, so they must be adjusted, and the ancestor class needs to be changed for the `ITK_EXERCISE_BASIC_OBJECT_METHODS` for the `itkIntermodesThresholdImageFilter` tests), it can be checked that the tests affected by this issue (e.g. `itkIntermodesThresholdImageFilterTest`, `itkIntermodesThresholdImageFilterTestNoInterMode`, `itkIntermodesThresholdImageFilterTestShort`) pass.

I tried to avoid the duplicated code in the `GenerateData` method, but since the `itk::Statistics::ImageToHistogramFilter` and the `itk::Statistics::MaskedImageToHistogramFilter` classes do not hae an immediate common ancestor, I did not quite find the way to do it, so unless there is some proposal to do that, and if CI builds turn out to be successful, this is ready to be merged.

I will rebase PR #511 on top of `master` once this gets merged, and address the `m_Min`/`m_Max` ivar initialization issue in a separate PR to again rebase PR #511 on top of `master` once that gets solved.

Reviews are welcome.

blowekamp commented on Modules/Filtering/Thresholding/include/itkHistogramThresholdImageFilter.h:
```diff
@@ -59,22 +50,23 @@ namespace itk
 template <typename TInputImage, typename TOutputImage, typename TMaskImage = TOutputImage>
-class ITK_TEMPLATE_EXPORT HistogramThresholdImageFilter : public ImageToImageFilter<TInputImage, TOutputImage>
+class ITK_TEMPLATE_EXPORT HistogramThresholdImageFilter
+  : public HistogramThresholdImageFilterBase<TInputImage, TOutputImage>
```
Shouldn't the `TMaskImage` template parameter be passes to the parent class.

jhlegarreta replied:
My bad. Thanks Brad. Done in 624c503.

blowekamp commented on Modules/Filtering/Thresholding/include/itkHistogramThresholdImageFilterBase.h:
```diff
@@ -0,0 +1,202 @@
+ * Melbourne, Australia.
+ * \author Gaetan Lehmann. Biologie du Developpement et de la Reproduction, INRA de Jouy-en-Josas, France.
+ *
+ * This implementation was taken from the Insight Journal paper:
```
"This implementation was based on the Insight Journal paper:"

It also looks like you are authoring this class :)

jhlegarreta replied:
Thanks for considering it Brad, but I think the PR/commit authorship is enough credit.

blowekamp commented on Modules/Filtering/Thresholding/include/itkIntermodesThresholdImageFilter.h:
```diff
@@ -90,61 +90,57 @@ class IntermodesThresholdImageFilter : public HistogramThresholdImageFilter<TInp
 
+  /** Set/Get the calculator to use to compute the threshold */
+  itkSetObjectMacro(Calculator, CalculatorType);
+  itkGetModifiableObjectMacro(Calculator, CalculatorType);
```
The parent class has a public Set/Get Calculator method. This method should be overloaded. The set method likely needs a dynamic_cast of the calculator.

blowekamp replied:
Consider the alternative solution of overloading these methods so that HistogramThresholdImageFilter::m_Calclulcator and this::m_ItermodeCalculator point to the same instance.

The overloaded HistogramThresholdImageFilter::SetCalculator method would need to dynamic_cast to the Intermodes calculator, the call HistogramThresholdImageFilter::SetCalculator and this::SetCalculator with the dynamic casted object.

jhlegarreta replied:
Not sure how you would create the `m_Calculator` object in [here](https://github.com/InsightSoftwareConsortium/ITK/pull/1718/files#diff-a4eb676b8dccbc553855aabae3ec343fR34), though

blowekamp replied:
HistogramThresholdImageFilter still have its m_Calculator ivar.
IntermodesThresholdImageFilter maintains its original m_IntermodesCalculator with derived type.

Both SetCalculator methods would set both variables.

jhlegarreta replied:
Thanks for the comments Brad. Please check commit 624c503. Especially the change in [here](https://github.com/InsightSoftwareConsortium/ITK/pull/1718/files#diff-e8fefc8c747bbff3a35dbb5c9fbfe84eL116) as I did not have a clear criterion to continue using an `itkGetModifiableObjectMacro` macro, and the base class `GetCalculator` implementation in [here](https://github.com/InsightSoftwareConsortium/ITK/pull/1718/files#diff-acc13746dfc9bfba2ffc93d33947b390R186).

blowekamp commented on Modules/Filtering/Thresholding/include/itkIntermodesThresholdImageFilter.hxx:
```diff
@@ -0,0 +1,159 @@
+IntermodesThresholdImageFilter<TInputImage, TOutputImage, TMaskImage>::SetMaximumSmoothingIterations(
+  SizeValueType maxSmoothingIterations)
+{
+  m_Calculator->SetMaximumSmoothingIterations(maxSmoothingIterations);
```
Does this cause the modify time of the image filter class to be updated?

The simple way to do this would be to copy the code in the set/get macro which check if the value is the same, if they are different then set the modify time. The more complicated way would involve modify the parent's GetMTime method to return the max of the MTime of itself, and the m_Calculator.

jhlegarreta replied:
Opened issue #1720.

blowekamp commented on Modules/Filtering/Thresholding/include/itkIntermodesThresholdImageFilter.hxx:
```diff
@@ -0,0 +1,159 @@
+{
+  if (m_Calculator.IsNull())
+  {
+    itkExceptionMacro(<< "No threshold calculator set.");
```
This type of check should happen the the "VerifyPreconditions" method.

jhlegarreta replied:
Done in 624c503.

blowekamp commented:
Looks like a very nice refactor to a base class. However it duplicates code into IntermodesThresholdImageFilter::Gernate data. Would it be simpler to overload the Set/Get Calculator methods to set the calculator ivars in the parent and Intermodes filter to be the same calculator  object ( although in HistogramThresholdImageFilter would point to the parent class, while Intermodes would point to he derived. )?

blowekamp commented:
May I take a stab at addressing the issue, as I don't believe the base class refactor is nessacary to address?

dzenanz commented:
If I remember correctly, he was looking for help 😄

jhlegarreta commented:
> If I remember correctly, he was looking for help 😄

@dzenanz Brad made a few comments early on and following them I was able to propose a solution that was successful.

> May I take a stab at addressing the issue, as I don't believe the base class refactor is nessacary to address?

@blowekamp I thought that after having addressed the comments the solution I came up with was a pretty neat one, and had tested with the failing tests in #511, but definitely, if you think you have a more elegant or efficient one, please go ahead.

blowekamp commented:
Ok, hopefully it works out elegantly... I also have interest in making these filters run "inplace" and the extra base class would make it more difficult.

jhlegarreta commented:
Superseded by #1730. Closing.
